modules/ingester/flush.go

```diff
@@ -124,16 +179,46 @@ func (i *Ingester) flushLoop(j int) {
 		}
 		op := o.(*flushOp)
 
-		level.Debug(log.Logger).Log("msg", "flushing block", "userid", op.userID, "fp")
-
-		err := i.flushUserTraces(op.userID)
-		if err != nil {
-			level.Error(log.WithUserID(op.userID, log.Logger)).Log("msg", "failed to flush user", "err", err)
-
-			// re-queue failed flush
-			op.from += int64(flushBackoff)
-			i.flushQueues.Requeue(op)
-			continue
+		if op.kind == complete {
+			level.Debug(log.Logger).Log("msg", "completing block", "userid", op.userID, "fp")
+			instance, exists := i.getInstanceByID(op.userID)
+			if !exists {
+				// instance no longer exists? that's bad, clear and continue
+				_ = op.completingBlock.Clear()
+				continue
+			}
+
+			completeBlock, err := instance.writer.CompleteBlock(op.completingBlock, instance)
+
+			instance.blocksMtx.Lock()
+			if err != nil {
+				// this is a really bad error that results in data loss.  most likely due to disk full
+				_ = op.completingBlock.Clear()
+				metricFailedFlushes.Inc()
+				level.Error(log.Logger).Log("msg", "unable to complete block.  THIS BLOCK WAS LOST", "tenantID", op.userID, "err", err)
+				instance.blocksMtx.Unlock()
+				continue
+			}
+			instance.completeBlocks = append(instance.completeBlocks, completeBlock)
+			instance.blocksMtx.Unlock()
+		} else {
+			level.Debug(log.Logger).Log("msg", "flushing block", "userid", op.userID, "fp")
+
+			err := i.flushUserTraces(op.userID)
+			if err != nil {
+				level.Error(log.WithUserID(op.userID, log.Logger)).Log("msg", "failed to flush user", "err", err)
+
```

Comment:
Perhaps it's interesting to increment metricFailedFlushes here.

Reply:
Right, that is interesting. I thought about that but since we're re-queuing the op, it made sense to metric on the "bad" errors that cannot be retried.